File: mobile/patch_tp.js
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'node_modules', 'react-native-track-player', 'android', 'src', 'main', 'java', 'com', 'doublesymmetry', 'trackplayer', 'module', 'MusicModule.kt');

let content = fs.readFileSync(filePath, 'utf8');

let lines = content.split('\n');
let newLines = [];
let braceCount = 0;
let insideTarget = false;

for (let i = 0; i < lines.length; i++) {
    let line = lines[i];

    // If it was already patched but missing the closing brace, we still need to fix it.
    // Or if it wasn't patched yet.
    if (line.includes('= scope.launch {')) {
        line = line.replace('= scope.launch {', '{ scope.launch {');
        insideTarget = true;
        braceCount = 0;
    } else if (line.includes('{ scope.launch {') && !line.includes('//')) {
        insideTarget = true;
        braceCount = 0;
    }

    newLines.push(line);

    if (insideTarget) {
        for (let j = 0; j < line.length; j++) {
            if (line[j] === '{') braceCount++;
            if (line[j] === '}') braceCount--;
        }

        // The outer function adds 1 to braceCount, and scope.launch adds another 1.
        // When we hit the end of scope.launch, braceCount will drop to 1.
        // And the line is exactly "    }"
        if (braceCount === 1 && line.trim() === '}') {
            newLines.push('    }'); // Close the outer function
            insideTarget = false;
        }
    }
}

fs.writeFileSync(filePath, newLines.join('\n'));
console.log('Successfully fully patched MusicModule.kt');

File: apps/mobile/patch_tp.py
import os

file_path = os.path.join("node_modules", "react-native-track-player", "android", "src", "main", "java", "com", "doublesymmetry", "trackplayer", "module", "MusicModule.kt")

with open(file_path, "r", encoding="utf-8") as f:
    content = f.read()

chunks = content.split("\n    @ReactMethod\n")

new_chunks = []
for i, chunk in enumerate(chunks):
    # Fix the `= scope.launch {`
    if ") = scope.launch {" in chunk:
        chunk = chunk.replace(") = scope.launch {", ") { scope.launch {")
        if i == len(chunks) - 1:
            # Last chunk has the class closing brace `\n}` at the end
            chunk = chunk.rstrip()
            if chunk.endswith("}"):
                chunk = chunk[:-1] + "    }\n}"
        else:
            # Append the closing brace for the method
            chunk = chunk.rstrip() + "\n    }\n"
    elif ") =\n        scope.launch {" in chunk:
        chunk = chunk.replace(") =\n        scope.launch {", ") {\n        scope.launch {")
        if i == len(chunks) - 1:
            chunk = chunk.rstrip()
            if chunk.endswith("}"):
                chunk = chunk[:-1] + "    }\n}"
        else:
            chunk = chunk.rstrip() + "\n    }\n"
    
    new_chunks.append(chunk)

new_content = "\n    @ReactMethod\n".join(new_chunks)

# Fix the originalItem nullability issue for Arguments.fromBundle
new_content = new_content.replace("musicService.tracks[index].originalItem)", "musicService.tracks[index].originalItem ?: Bundle())")
new_content = new_content.replace("musicService.tracks[musicService.getCurrentTrackIndex()].originalItem\n            )", "musicService.tracks[musicService.getCurrentTrackIndex()].originalItem ?: Bundle()\n            )")
new_content = new_content.replace("musicService.tracks.map { it.originalItem }", "musicService.tracks.map { it.originalItem ?: Bundle() }")

with open(file_path, "w", encoding="utf-8") as f:
    f.write(new_content)

print("Successfully patched MusicModule.kt")
